main.go

```diff
@@ -1014,58 +1022,137 @@ func processImportCfg(flags []string) (newImportCfg string, _ error) {
 	return newCfg.Name(), nil
 }
 
-type knownReflect struct {
-	pkgPath string
-	name    string
-}
-
 type funcFullName = string
+type reflectParameterPosition = int
 
 // knownReflectAPIs is a static record of what std APIs use reflection on their
 // parameters, so we can avoid obfuscating types used with them.
 //
-// For now, this table is manually maintained, until we automatically detect and
-// propagate the uses of reflect in std and third party APIs.
-// If we end up maintaining this table for a long time, we should probably
-// improve the process via either code generation or sanity checks.
-//
-// TODO: record which parameters get used with reflection, as it's often not all
-// of them.
-//
 // TODO: we're not including fmt.Printf, as it would have many false positives,
 // unless we were smart enough to detect which arguments get used as %#v or %T.
-//
-// TODO: this should also support third party APIs; see
-// https://github.com/burrowers/garble/issues/162
-var knownReflectAPIs = map[funcFullName]bool{
-	"reflect.TypeOf":  true,
-	"reflect.ValueOf": true,
-
-	"encoding/json.Marshal":           true,
-	"encoding/json.MarshalIndent":     true,
-	"encoding/json.Unmarshal":         true,
-	"(*encoding/json.Decoder).Decode": true,
-	"(*encoding/json.Encoder).Encode": true,
-
-	"encoding/gob.Register":          true,
-	"encoding/gob.RegisterName":      true,
-	"(*encoding/gob.Decoder).Decode": true,
-	"(*encoding/gob.Encoder).Encode": true,
-
-	"encoding/xml.Marshal":                  true,
-	"encoding/xml.MarshalIndent":            true,
-	"encoding/xml.Unmarshal":                true,
-	"(*encoding/xml.Decoder).Decode":        true,
-	"(*encoding/xml.Encoder).Encode":        true,
-	"(*encoding/xml.Decoder).DecodeElement": true,
-	"(*encoding/xml.Encoder).EncodeElement": true,
-
-	"(*text/template.Template).Execute": true,
-
-	"(*html/template.Template).Execute": true,
-
-	"net/rpc.Register":     true,
-	"net/rpc.RegisterName": true,
+var knownReflectAPIs = map[funcFullName][]reflectParameterPosition{
+	"reflect.TypeOf":  {0},
+	"reflect.ValueOf": {0},
+}
+
+func loadKnownReflectAPIs(currPkg *listedPackage) error {
+	for path := range importCfgEntries {
+		pkg, err := listPackage(path)
+		if err != nil {
+			return err
+		}
+
+		err = func() error {
+			f, err := os.Open(pkg.Export + "-garble")
+			if err != nil {
+				return err
+			}
+			defer f.Close()
+
+			var newKnownReflectAPIs map[string][]int
+			if err := gob.NewDecoder(f).Decode(&newKnownReflectAPIs); err != nil {
```

Comment:
You can just decode straight into `knownReflectAPIs` and it should just work by merging entries:

> In general, if allocation is required, the decoder will allocate memory. If not, it will update the destination variables with values read from the stream. It does not initialize them first, so if the destination is a compound value such as a map, struct, or slice, the decoded values will be merged elementwise into the existing variables.

Reply:
Intresting I didn't know that